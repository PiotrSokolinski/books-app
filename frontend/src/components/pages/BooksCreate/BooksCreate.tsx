import { Create, ShowButton, useTranslate } from 'react-admin';

import { BookActions } from '../../organisms/BookActions/BookActions';
import { BookForm } from '../../organisms/BookForm/BookForm';
import { Wrapper } from '../../templates/Wrapper/Wrapper';

export const BooksCreate = () => {
  const translate = useTranslate();

  return (
    <Wrapper>
      <Create
        title={translate('booksCreate.barTitle')}
        actions={<BookActions action={<ShowButton />} />}
      >
        <BookForm mode="create" title={translate('booksCreate.formTitle')} />
      </Create>
    </Wrapper>
  );
};
