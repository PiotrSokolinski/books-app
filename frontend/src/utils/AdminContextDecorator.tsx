import React from 'react';
import {
  AdminContext,
  DataProvider,
  GetListResult,
  memoryStore,
  testDataProvider,
} from 'react-admin';

import { theme } from '../theme/theme';
import { i18nProvider } from '../i18n/i18n';
import { Book } from '../types/book.ts';

const defaultDataProvider: DataProvider = testDataProvider({
  getList: async (): Promise<GetListResult<Book>> => {
    const books: Book[] = [
      {
        id: '1',
        title: 'The Hobbit',
        author: 'J.R.R. Tolkien',
        isbnNumber: '1234567890',
        numberOfPages: 300,
        rating: 4,
      },
      {
        id: '2',
        title: 'The Lord of the Rings',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    ];

    return Promise.resolve({
      total: books.length,
      data: books,
    });
  },
  getOne: () => {
    return Promise.resolve({
      data: {
        id: '1',
        title: 'The Lord of the Rings',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  create: () => {
    return Promise.resolve({
      data: {
        id: '3',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  update: () => {
    return Promise.resolve({
      data: {
        id: '2',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
  delete: () => {
    return Promise.resolve({
      data: {
        id: '3',
        title: 'Test Book Title',
        author: 'J.R.R. Tolkien',
        isbnNumber: '0987654321',
        numberOfPages: 1000,
        rating: 5,
      },
    });
  },
} as unknown as DataProvider); // https://github.com/marmelab/react-admin/issues/5476

// TODO: F7 - Add more documentation and comments to the codebase
/**
 * A decorator component that provides the React-Admin context for testing or storybook environments.
 *
 * This component wraps its children with the necessary React-Admin context providers,
 * including a theme, internationalization provider, and optionally a `dataProvider`.
 * The `authProvider` is set to `undefined`, as this provider is designed for simple use cases.
 *
 * It supports passing a custom `dataProvider` for testing or storybook purposes.
 *
 * @param {Object} props - The props for the component.
 * @param {React.ReactNode} props.children - The child components to render within the React-Admin context.
 * @param {Partial<DataProvider>} [props.dataProvider] - An optional custom `dataProvider` to be used in place of the default one.
 *
 * @returns {JSX.Element} The wrapped children within the React-Admin context, with context providers like theme and internationalization applied.
 *
 * @example
 * <AdminContextDecorator>
 *   <MyComponent />
 * </AdminContextDecorator>
 */
export const AdminContextDecorator: React.FC<{
  children: React.ReactNode;
  dataProvider?: Partial<DataProvider>;
}> = ({
  children,
  dataProvider,
}: {
  children: React.ReactNode;
  dataProvider?: Partial<DataProvider>;
}): React.JSX.Element => (
  <AdminContext
    dataProvider={
      dataProvider ? testDataProvider(dataProvider) : defaultDataProvider
    }
    authProvider={undefined}
    theme={theme}
    store={memoryStore()}
    i18nProvider={i18nProvider}
  >
    {children}
  </AdminContext>
);
