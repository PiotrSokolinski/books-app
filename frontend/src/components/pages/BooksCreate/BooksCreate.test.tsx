import { render, screen } from '@testing-library/react';
import { vi, describe, it, expect } from 'vitest';
import { fireEvent } from '@testing-library/react';
import { ResourceContextProvider } from 'react-admin';

import { BooksCreate } from './BooksCreate';
import { AdminContextDecorator } from '../../../utils/AdminContextDecorator.tsx';

vi.doMock(import('react-admin'), async (importOriginal) => {
  const actual = await importOriginal();

  return {
    ...actual,
    useTranslate: vi.fn().mockReturnValue((key: string) => key),
    useRedirect: vi.fn().mockReturnValue(() => vi.fn()),
  };
});

vi.mock(import('react-hook-form'), async (importOriginal) => {
  const actual = await importOriginal();

  return {
    ...actual,
    useFormContext: vi.fn().mockReturnValue({
      reset: vi.fn(),
    }),
  };
});

// TODO: F2 - More integration tests across the application
describe('BooksCreate', () => {
  it('renders the BooksCreate component and submits a record', async () => {
    render(
      <AdminContextDecorator>
        <ResourceContextProvider value="books">
          <BooksCreate />
        </ResourceContextProvider>
      </AdminContextDecorator>,
    );

    fireEvent.change(screen.getByLabelText(/title/i), {
      target: { value: 'New Book Title' },
    });
    fireEvent.change(screen.getByLabelText(/author/i), {
      target: { value: 'Author Name' },
    });
    fireEvent.change(screen.getByLabelText(/isbn number/i), {
      target: { value: '1234567890' },
    });
    fireEvent.change(screen.getByLabelText(/number of pages/i), {
      target: { value: '200' },
    });
    fireEvent.change(screen.getByLabelText(/rating/i), {
      target: { value: '4.5' },
    });

    const saveButton = screen.getByText(/save/i);
    expect(saveButton).toBeInTheDocument();

    fireEvent.click(saveButton);
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <AdminContextDecorator>
        <ResourceContextProvider value="books">
          <BooksCreate />
        </ResourceContextProvider>
      </AdminContextDecorator>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
