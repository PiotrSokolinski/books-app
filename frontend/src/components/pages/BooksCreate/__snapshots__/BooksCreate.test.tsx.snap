// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BooksCreate > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1es8s0n"
  >
    <div
      class="create-page css-ptxxna-RaCreate-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-dense css-8czsxw-MuiToolbar-root-RaTopToolbar-root"
      >
        <div
          class="MuiBox-root css-1oxffa5"
        >
          <button
            aria-label="Back to list"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-7nabzj-MuiButtonBase-root-MuiButton-root-RaButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeSmall css-y6rp3m-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowBackIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20z"
                />
              </svg>
            </span>
            Back to list
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="RaCreate-main"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root RaCreate-card css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <form>
            <div
              class="MuiCardContent-root css-1fd8x6x-MuiCardContent-root-RaSimpleForm-root"
            >
              <div
                class="MuiStack-root css-1a69w1n-MuiStack-root"
              >
                <h5
                  class="MuiTypography-root MuiTypography-h5 css-1iranwn-MuiTypography-root"
                >
                  Add a new book
                </h5>
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-title css-1i5wpaj-MuiFormControl-root-MuiTextField-root-RaResettableTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":ra:"
                    id=":ra:-label"
                  >
                    <span>
                      Title
                      <span
                        aria-hidden="true"
                      >
                         *
                      </span>
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1lm524j-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-describedby=":ra:-helper-text"
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1zuho2-MuiInputBase-input-MuiFilledInput-input"
                      id=":ra:"
                      name="title"
                      type="text"
                      value=""
                    />
                  </div>
                  <p
                    class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
                    id=":ra:-helper-text"
                  >
                    <span>
                      ​
                    </span>
                  </p>
                </div>
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-author css-1i5wpaj-MuiFormControl-root-MuiTextField-root-RaResettableTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":rc:"
                    id=":rc:-label"
                  >
                    <span>
                      Author
                      <span
                        aria-hidden="true"
                      >
                         *
                      </span>
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1lm524j-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-describedby=":rc:-helper-text"
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1zuho2-MuiInputBase-input-MuiFilledInput-input"
                      id=":rc:"
                      name="author"
                      type="text"
                      value=""
                    />
                  </div>
                  <p
                    class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
                    id=":rc:-helper-text"
                  >
                    <span>
                      ​
                    </span>
                  </p>
                </div>
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-isbnNumber css-1i5wpaj-MuiFormControl-root-MuiTextField-root-RaResettableTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":re:"
                    id=":re:-label"
                  >
                    <span>
                      ISBN number
                      <span
                        aria-hidden="true"
                      >
                         *
                      </span>
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1lm524j-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-describedby=":re:-helper-text"
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1zuho2-MuiInputBase-input-MuiFilledInput-input"
                      id=":re:"
                      name="isbnNumber"
                      type="text"
                      value=""
                    />
                  </div>
                  <p
                    class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
                    id=":re:-helper-text"
                  >
                    <span>
                      ​
                    </span>
                  </p>
                </div>
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-numberOfPages css-m4dhox-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":rg:"
                    id=":rg:-label"
                  >
                    <span>
                      Number of pages
                      <span
                        aria-hidden="true"
                      >
                         *
                      </span>
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1lm524j-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-describedby=":rg:-helper-text"
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1zuho2-MuiInputBase-input-MuiFilledInput-input"
                      id=":rg:"
                      name="numberOfPages"
                      step="any"
                      type="number"
                      value=""
                    />
                  </div>
                  <p
                    class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
                    id=":rg:-helper-text"
                  >
                    <span>
                      ​
                    </span>
                  </p>
                </div>
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root ra-input ra-input-rating css-m4dhox-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-filled css-1mog9k0-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":ri:"
                    id=":ri:-label"
                  >
                    <span>
                      Rating
                      <span
                        aria-hidden="true"
                      >
                         *
                      </span>
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1lm524j-MuiInputBase-root-MuiFilledInput-root"
                  >
                    <input
                      aria-describedby=":ri:-helper-text"
                      aria-invalid="false"
                      class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-1zuho2-MuiInputBase-input-MuiFilledInput-input"
                      id=":ri:"
                      name="rating"
                      step="any"
                      type="number"
                      value=""
                    />
                  </div>
                  <p
                    class="MuiFormHelperText-root MuiFormHelperText-sizeSmall MuiFormHelperText-contained css-k4qjio-MuiFormHelperText-root"
                    id=":ri:-helper-text"
                  >
                    <span>
                      ​
                    </span>
                  </p>
                </div>
              </div>
            </div>
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1iw8x7s-MuiToolbar-root"
            >
              <button
                aria-label="Save"
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1f0c020-MuiButtonBase-root-MuiButton-root-RaSaveButton-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="SaveIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3m3-10H5V5h10z"
                  />
                </svg>
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
